qubrabench/algorithms/search.py

```diff
@@ -122,6 +124,97 @@ def search(
     return None
 
 
+class SamplingDomain(ABC, Generic[E]):
+    """
+    Class used for search routines requiring elaborate sampling with replacement. To search against a predicate, please
+    refer to the function search_custom.
+    If your use-case involves simpler search spaces, e.g. ones that can be stored as a list in memory, please use the
+    function search for that.
+    """
+
+    @abstractmethod
+    def get_size(self) -> int:
+        pass
+
+    @abstractmethod
+    def get_probability_of_sampling_solution(self, key) -> float:
+        pass
+
+    @abstractmethod
+    def get_random_sample(self, rng: np.random.Generator) -> E:
+        pass
+
+
+def search_by_sampling_with_replacement(
+    domain: SamplingDomain[E],
+    key: Callable[[E], bool],
+    *,
+    rng: np.random.Generator,
+    error: float,
+    max_classical_queries: int = 130,
+) -> Optional[E]:
+    """Search a domain by sampling with replacement, for an element satisfying the given predicate, while keeping track of query statistics.
+
```

Comment:
explain how the costs are computed exactly

Reply:
tracked in #120